### PR TITLE
feat(core/plugins): Plugins redesign 

### PR DESCRIPTION
**Pull request by peri044**
# Description

Redesigning plugin API according to https://github.com/NVIDIA/TRTorch/discussions/385

Changes all plugins mentioned in https://github.com/NVIDIA/TRTorch/discussions/385

## Type of change

Please delete options that are not relevant and/or add your own.

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [x] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: core/plugins/impl/normalize_plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+  cudaEvent_t torch_event;
+  cudaEventCreate(&torch_event);
+  cudaEventRecord(torch_event, torch_stream.stream());
+
+  cudaStreamWaitEvent(stream, torch_event, 0);
+
+  cudaEventDestroy(event);
+  cudaEventDestroy(torch_event);
+  return 0;
+#else
```
Comment (narendasan) [inline core/plugins/impl/normalize_plugin.cpp:208]: Are we still stuck with CPU execution for ATen?

File: core/plugins/register_plugins.cpp [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+class TRTorchPluginRegistry {
+ public:
+  TRTorchPluginRegistry() {
+    auto trtorch_logger = util::logging::TRTorchLogger(
+        "[TRTorch Plugins Context] - ",
+        util::logging::get_logger().get_reportable_severity(),
+        util::logging::get_logger().get_is_colored_output_on());
+    // register libNvInferPlugins and TRTorch plugins
+    initLibNvInferPlugins(&trtorch_logger, "");
+    REGISTER_TRTORCH_PLUGIN(InterpolatePluginCreator);
```
Comment (narendasan) [inline core/plugins/register_plugins.cpp:23]: Can we just have the Register calls at the bottom of their respective source files? I think we only need the static object for the logger and the nvinfer library. Also we should probably hold the trtorch_logger as a member variable so it only gets destroyed at the end.

File: examples/sample_rt_app/README.md [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+ 2) Build and run `samplertapp`
+
+​     `samplertapp` is a binary which loads the torchscript modules `conv_gelu.jit` or `norm.jit` and runs the TRT engines on a random input using TRTorch runtime components. Checkout the `main.cpp` and `BUILD ` file for necessary code and compilation dependencies.
+
+To build and run the app
+
+```
+cd TRTorch
+bs run //examples/sample_rt_app:samplertapp $PWD/examples/sample_rt_app/norm.jit
```
Comment (narendasan) [inline examples/sample_rt_app/README.md:35]: is bs a valid alias for bazel?

File: examples/sample_rt_app/main.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#include <iostream>
+#include <fstream>
+#include <memory>
+#include <sstream>
+#include <vector>
+#include "torch/script.h"
+
+// Load the TRT engine from engine_path
+std::vector<char> loadEngine(std::string engine_path){
```
Comment (narendasan) [inline examples/sample_rt_app/main.cpp:9]: What is this function used for?

File: core/conversion/var/Var.cpp [hunk 5]
```diff
@@ -100,8 +100,10 @@ nvinfer1::ITensor* Var::ITensorOrFreeze(ConversionCtx* ctx) {
     if (ptr_.ivalue->isTensor()) {
       auto weights = converters::Weights();
       auto tensor = ptr_.ivalue->toTensor();
-      if ((tensor.scalar_type() == at::kLong || tensor.scalar_type() == at::kDouble) && !ctx->settings.truncate_long_and_double) {
-        TRTORCH_THROW_ERROR("Unable to freeze tensor of type Int64/Float64 into constant layer, try to compile model with truncate_long_and_double enabled");
+      if ((tensor.scalar_type() == at::kLong || tensor.scalar_type() == at::kDouble) &&
```
[the remark this answers is not in the document]
Comment (peri044) [reply core/conversion/var/Var.cpp:103]: Probably a linter change. I did not make any custom changes to this file

File: core/plugins/impl/normalize_plugin.cpp (hunk 1 above)
Comment (peri044) [reply core/plugins/impl/normalize_plugin.cpp:208]: I did not try with GPU execution yet.

File: core/plugins/register_plugins.cpp (hunk 2 above)
Comment (peri044) [reply core/plugins/register_plugins.cpp:23]: I'll move the logger object as a member variable. 
When I compiled the REGISTER calls with their source files, the `libtrtorchrt` was unable to find the plugin creators. When I moved it to this `register_plugins.cpp`, I could see them getting registered correctly. I shall double check.

File: examples/sample_rt_app/README.md (hunk 3 above)
Comment (peri044) [reply examples/sample_rt_app/README.md:35]: Fixed it.

File: examples/sample_rt_app/main.cpp (hunk 4 above)
Comment (peri044) [reply examples/sample_rt_app/main.cpp:9]: Removed it now.

File: core/plugins/plugins.h [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+namespace trtorch {
+namespace core {
+namespace plugins {
+namespace impl {
+// Helper class which registers a plugin in trtorch namespace
+template <typename T>
+class PluginRegistrar {
+ public:
+  PluginRegistrar() {
+    getPluginRegistry()->registerCreator(instance, "trtorch");
```
Comment (narendasan) [inline core/plugins/plugins.h:16]: add nvinfer1:: here to be clear

Comment (narendasan) [inline core/plugins/plugins.h:16]: Add a LOG_DEBUG Saying what Plugin is being registered in the TensorRT Plugin registry

Comment (peri044) [reply core/plugins/plugins.h:16]: Modified this to include more debug information on what plugins are being registered.

File: core/plugins/register_plugins.cpp (hunk 2 above)
Comment (peri044) [reply core/plugins/register_plugins.cpp]: I moved it to respective source files and tested it and it seemed to work now to my surprise. When I tested it a couple of days back, it did not. I'd appreciate if you can try this out on your system just to validate the behavior. The way I currently test this is by building TRTorch tar file and running `sample_rt_app`

File: core/plugins/plugins.h (hunk 6 above)
Comment (peri044) [reply core/plugins/plugins.h:16]: Just saw `getPluginRegistry` is not in any namespace.

**Comment by narendasan** [comment]
@peri044 can you rebase on master and fix linting. There were some things i fixed in another pr

**Comment by narendasan** [comment]
Code wise LGTM

**Comment by peri044** [comment]
@narendasan Fixed the linting issues.

**Comment by narendasan** [comment]
@peri044 can you rebase and fix any conflicts by EOD tomorrow?

**Comment by peri044** [comment]
@narendasan done. Good to merge now

**Comment by peri044** [comment]
@narendasan Rebased it.

**Review by narendasan — APPROVED** [review]
LGTM

**Pull request merged**